### DISCUSSION TITLE
Issues after encountering issues with netbox instance

### DISCUSSION THREAD

**Discussion by ijjoseph** [netbox-community/netbox#8991] [Q&A]
Hi All, 

I need some assistance with recovering a database.
See below error message received. I'm not sure where to begin. Can someone help?

**Server Error**
A database programming error was detected while processing this request. Common causes include the following:

**Database migrations missing** - When upgrading to a new NetBox release, the upgrade script must be run to apply any new database migrations. You can run migrations manually by executing python3 manage.py migrate from the command line.

**Unsupported PostgreSQL version** - Ensure that PostgreSQL version 9.6 or higher is in use. You can check this by connecting to the database using NetBox's credentials and issuing a query for SELECT VERSION().

The complete exception is provided below:

<class 'django.db.utils.ProgrammingError'>

column extras_customfield.created does not exist
LINE 1: SELECT "extras_customfield"."created", "extras_customfield"....
               ^

Python version: 3.8.10
NetBox version: 3.1.10

![image](https://user-images.githubusercontent.com/42877086/160658296-fe2687ee-fc71-4f3e-b615-1c01eef9d730.png)

**Comment by jeremystretch** [comment]
The database error indicates that your migrations have not been run. What action did you take prior to receiving this error? Was NetBox recently upgraded?

**Reply by ijjoseph** [reply]
Hi Jeremy, I ran the upgrade script as per the instructions found here "https://netbox.readthedocs.io/en/stable/installation/upgrading/"

There were no errors displayed when i ran the script. should i run it again?

How do i run this migration?

**Reply by ijjoseph** [reply]
Hi, @jeremystretch is it possible to revert to the previous version of NetBox? We had an issue with a NetBox installation that was running the version stated in the screenshot below. I had neglected to save or backup the installation before I proceeded to update to the latest version via cloning from the git repo. 
How can I salvage this.? 
What steps are recommended to restore this install and bring it up to date?

<img width="609" alt="Screen Shot 2022-03-29 at 10 30 05 AM" src="https://user-images.githubusercontent.com/42877086/161097336-891341dc-b7c3-45de-b8e7-b076dcb45dc0.png">

**Reply by ijjoseph** [reply]
This is the error message I got when I ran manage.py migrate

/opt/netbox/netbox$ python3 manage.py migrate
/opt/netbox/netbox/netbox/settings.py:57: UserWarning: The CACHE_TIMEOUT configuration parameter was removed in v3.0.0 and no longer has any effect.
  warnings.warn(
/opt/netbox/netbox/netbox/settings.py:61: UserWarning: The RELEASE_CHECK_TIMEOUT configuration parameter was removed in v3.0.0 and no longer has any effect.
  warnings.warn(
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.8/dist-packages/django/core/management/__init__.py", line 377, in execute
    django.setup()
  File "/usr/local/lib/python3.8/dist-packages/django/__init__.py", line 24, in setup
    apps.populate(settings.INSTALLED_APPS)
  File "/usr/local/lib/python3.8/dist-packages/django/apps/registry.py", line 91, in populate
    app_config = AppConfig.create(entry)
  File "/usr/local/lib/python3.8/dist-packages/django/apps/config.py", line 90, in create
    module = import_module(entry)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'graphiql_debug_toolbar'

**Reply by samvdl** [reply]
Hi ijjoseph,

I encountered exactly the same situation.
First I wanted to upgrade from v2.8.9 to v3.2,5 which horribly failed on this issue
Than I read somewhere that you can't do a major upgrade in one time and you need to at least go by a minor one.
So I reverted / rolled back my VM and tried a minor one (v2.9.10), and I encountered the same issue.
So somewhere deep in the discussions on the net, there was this tread suggesting to start fiddling with the postgre database tables itself, but i'd pass for that (already deep enough in 'wonderland' as it is). The other suggestion was that you can's skip any minor upgrade...
So... again I reverted / rolled back my VM and tried the smallest minor upgrade (all patches = last number), so from V2.8.9 to V2.9.11. And yet again same error.
My PostgreSQL version is checked by Select version() and showing me PostgreSQL 11.16
applying Python3 /opt/netbox/netbox/manage.py migrate (knowing that /opt/netbox is a symlink to /opt/netbox/netbox-2.9.11 --> which is the new netbox version)

You got any solutions found in the mean while?

Thanks,
Sam

**Reply by cpmills1975** [reply]
In prod/dev I have a daily migration happening between 2.10.3 and 3.2.5. The requirement as I understand it is that you can't leap between 2.10 and 3.anything, you need to go to 2.11.something first.

This works consistently.

**Reply by samvdl** [reply]
> In prod/dev I have a daily migration happening between 2.10.3 and 3.2.5. The requirement as I understand it is that you can't leap between 2.10 and 3.anything, you need to go to 2.11.something first.
> 
> This works consistently.

Well that's nice and possibly a whole other unrelated story all together.
The issue at hand which 'ijjoseph' initiated, is about the fact that whenever upgrading netbox, you get a database error altough your python version, your PostgreSQL version, etc is high enough according to the netbox upgrade documentation...
And furthermore, trying to perform "manage.py migrate" is generating the errors like his post showed (and mind for that matter).

So question is not if it's possible to upgrade , yes or no, but why we get stuck on the database migration...

**Reply by candlerb** [reply]
```
/usr/local/lib/python3.8/dist-packages/django/apps/config.py
```

That shows you're running the migrations wrongly.  You must run Netbox within its virtualenv - not with some random system-installed version of Django.  That is:

```
cd /opt/netbox
. venv/bin/activate
python netbox/manage.py migrate
```

However, if instead you run the supplied `upgrade.sh` then it takes care of that for you.

> is it possible to revert to the previous version of NetBox? We had an issue with a NetBox installation that was running the version stated in the screenshot below. I had neglected to save or backup the installation before I proceeded to update to the latest version via cloning from the git repo.
> How can I salvage this.?

No, it's not possible to move back from a newer version of Netbox to an older one, if you've already applied or partially-applied the migrations.  You have to restore from a backup.  If you don't have backups then your data is gone, unless you are able to sort out pressing forward to the new version.

**Comment by samvdl** [comment]
```
(venv) [root@netboxVM bin]# /usr/bin/python3.9 /opt/netbox-2.9.11/netbox/manage.py migrate
Traceback (most recent call last):
  File "/opt/netbox-2.9.11/netbox/manage.py", line 8, in <module>
    from django.core.management import execute_from_command_line
ModuleNotFoundError: No module named 'django'
```

**Reply by candlerb** [reply]
You're running the wrong version of python.  Given that `(venv)` is active, you should do

```
python netbox/manage.py ...
```
not
```
/usr/bin/python3.9 netbox/manage.py ...
```

(When the virtualenv is active, "python" is redirected to the correct version of python inside the virtualenv)

**Reply by samvdl** [reply]
_(When the virtualenv is active, "python" is redirected to the correct version of python inside the virtualenv)_

--> Well this might be the problem see: 
```
(venv) [root@netboxVM netbox-2.9.11]# python --version
Python 2.7.5
```
I need to redirect to the correct version inside this new venv... How?

and so...
```
(venv) [root@netboxVM netbox-2.9.11]# python netbox/manage.py migrate
Traceback (most recent call last):
  File "netbox/manage.py", line 8, in <module>
    from django.core.management import execute_from_command_line
ImportError: No module named django.core.management
```

**Comment by samvdl** [comment]
I used python 3 (-> points to 3.6) to create the new virtual environment, so I am actually uncertain why 'python version' in that virtual enviroment gives me Python-2.7.5
`[ansible@netboxVM ~]$ /usr/bin/pyhton3 -m venv /opt/netbox-2.9.11/venv`

**Comment by markkuleinio** [comment]
It always puzzles me how people get into such a corner that they **need** to run various commands manually instead of letting the documented `upgrade.sh` (with or without specifying the PYTHON variable) command deal with the details.

**Reply by samvdl** [reply]
Maybe this is why:
```
(venv) [root@netboxVM-2.9.11]# ll /usr/bin/python*
lrwxrwxrwx 1 root root     7 Apr 28  2020 /usr/bin/python -> python2
lrwxrwxrwx 1 root root     9 Apr 28  2020 /usr/bin/python2 -> python2.7
-rwxr-xr-x 1 root root  7144 Apr  2  2020 /usr/bin/python2.7
-rwxr-xr-x 1 root root  1835 Apr  2  2020 /usr/bin/python2.7-config
lrwxrwxrwx 1 root root    16 Apr 28  2020 /usr/bin/python2-config -> python2.7-config
lrwxrwxrwx 1 root root     9 Apr 28  2020 /usr/bin/python3 -> python3.6
-rwxr-xr-x 2 root root 11336 Apr  2  2020 /usr/bin/python3.6
lrwxrwxrwx 1 root root    17 Apr 28  2020 /usr/bin/python3.6-config -> python3.6m-config
-rwxr-xr-x 2 root root 11336 Apr  2  2020 /usr/bin/python3.6m
-rwxr-xr-x 1 root root   173 Apr  2  2020 /usr/bin/python3.6m-config
-rwxr-xr-x 1 root root  3403 Apr  2  2020 /usr/bin/python3.6m-x86_64-config
lrwxrwxrwx 1 root root    24 Jul  7 15:09 /usr/bin/python3.9 -> /usr/local/bin/python3.9
lrwxrwxrwx 1 root root    16 Apr 28  2020 /usr/bin/python3-config -> python3.6-config
lrwxrwxrwx 1 root root    14 Apr 28  2020 /usr/bin/python-config -> python2-config
```

**Reply by markkuleinio** [reply]
I don't see any problem there in that directory listing. Whatever `python` or `python3` commands happen inside `upgrade.sh` (except creating the venv) they happen inside the venv, not in system-global pythons. (And, the venv is created using whatever PYTHON is specified when running `upgrade.sh`, if the system-default `python3` is not the desired one.)

**Reply by samvdl** [reply]
> I don't see any problem there in that directory listing. Whatever `python` or `python3` commands happen inside `upgrade.sh` (except creating the venv) they happen inside the venv, not in system-global pythons. (And, the venv is created using whatever PYTHON is specified when running `upgrade.sh`, if the system-default `python3` is not the desired one.)

So yes, the venv is created with python3.6.. And I think for netbox upgrade stuff I need at least 3.8... So I'll deactivate and destroy the virtual env. I made an create an new one with the right Python version (/usr/bin/python3.9)...
and that's already for starters....

But still not clear to me why in this created env (with python v3.6) , the command `python --version`, i'm getting v2.7.5... so puzzled

**Reply by markkuleinio** [reply]
**You** made or `upgrade.sh` made new venv? Sorry if I sound harsh, I still don't know what is the problem that you cannot use `upgrade.sh`, maybe I missed some detail.

**Reply by candlerb** [reply]
> It always puzzles me how people get into such a corner that they **need** to run various commands manually instead of letting the documented `upgrade.sh` (with or without specifying the PYTHON variable) command deal with the details.

Yes, this.

I believe all the OP needs is a single command:

```
PYTHON=/usr/bin/python3.9 /opt/netbox/upgrade.sh
```

Then everything should work.  And if it gives an error, they should report the error output *from that exact command*.

If people decide to build their own virtualenvs or apply migrations without following the official instructions, we can't prevent them doing so - but they get to keep the pieces.

**Reply by samvdl** [reply]
> **You** made or `upgrade.sh` made new venv? Sorry if I sound harsh, I still don't know what is the problem that you cannot use `upgrade.sh`, maybe I missed some detail.

Yes, I made the new 'env' ... I created an 'venv' folder under the new /opt/netbox-x.y.z folder. Than  I ran `pyhton3.9 -m venv /opt/netbox-x.x.x/venv`. And finally I activated it by `source /opt/netbox-x.x.x/venv/bin/activate`. 
It was so advised by a colleague to always create a new environment for each upgrade.
I was not aware that 'upgrade.sh' creates a new environment.

Also important to mention (well maybe completely irrelevant as well) that I became conscious of the fact that my starting context actually is not the same as the context that is implicitly expected from within the upgrade documentation. This thanks to another contributor: [kkthxbye-code](https://github.com/kkthxbye-code).

So the original-first installed Netbox (v2.8.9) is natively installed in /opt/netbox on our server is actually NOT a symlink to the a sort-of reference folder called '.../netbox- 2.8.9/...' . But the documentation on upgrading expects this context.
So first of all, that was why I was completely not understanding the copy of the files.
I inherited this situation, so I did not brought myself in this situation. How could I have known?

Moreover I wonder if .upgrade.sh of the new netbox version is vulnerable to the context (meaning /opt/nebox being a symlink to the installation dir and not the native one). In the context of what the documentation is expecting and according to [kkthxbye-code] (a person ;-)), /opt/netbox just supposed to be a symlink to the a folder of a certain version.

I hope I'm making sense to you ;-)

**Reply by markkuleinio** [reply]
> Moreover I wonder if .upgrade.sh of the new netbox version is vulnerable to the context (meaning /opt/nebox being a symlink to the installation dir and not the native one).

That's not a problem, as you can see in the upgrade instructions: https://docs.netbox.dev/en/stable/installation/upgrading/

**Reply by samvdl** [reply]
> > Moreover I wonder if .upgrade.sh of the new netbox version is vulnerable to the context (meaning /opt/nebox being a symlink to the installation dir and not the native one).
> 
> That's not a problem, as you can see in the upgrade instructions: https://docs.netbox.dev/en/stable/installation/upgrading/

Well, it is when your native installation is in /opt/netbox/ and so when opt/netbox is NOT a symlink.
Imagine for 1 moment this to be the truth, than...
Than the documentation doesn't make any sense:
with X.Y.Z. being the new version:
sudo ln -sfn /opt/netbox-X.Y.Z/ /opt/netbox

is actually saying that /opt/netbox needs to change from a native installation directory towards a symlink pointing towards the new version directory.
Ofcourse this is nonsense... I see now.
So the documentation on upgrading is correct (well I pressume) assuming that the primordial installation was done accordingly...
And that's for starters, not the case with me (!) (ouch) I inherited it otherwise.
so I already now rectified this context towards the "meant to be" and restarted the netbox service. And that's good (working).
So now I can restart the upgrade (with I minor version) from a clean situation.
I'll keep you posted.

**Reply by markkuleinio** [reply]
> So now I can restart the upgrade (with I minor version) from a clean situation.

Yes: as long as **/opt/netbox** is the correct root path for your NetBox installation (whether a symlink or git clone or whatever), the install path is normal and the instructions apply as is.

**Reply by samvdl** [reply]
> > > Moreover I wonder if .upgrade.sh of the new netbox version is vulnerable to the context (meaning /opt/nebox being a symlink to the installation dir and not the native one).
> > 
> > 
> > That's not a problem, as you can see in the upgrade instructions: https://docs.netbox.dev/en/stable/installation/upgrading/
> 
> Well, it is when your native installation is in /opt/netbox/ and so when opt/netbox is NOT a symlink. Imagine for 1 moment this to be the truth, than... Than the documentation doesn't make any sense: with X.Y.Z. being the new version: sudo ln -sfn /opt/netbox-X.Y.Z/ /opt/netbox
> 
> is actually saying that /opt/netbox needs to change from a native installation directory towards a symlink pointing towards the new version directory. Ofcourse this is nonsense... I see now. So the documentation on upgrading is correct (well I pressume) assuming that the primordial installation was done accordingly... And that's for starters, not the case with me (!) (ouch) I inherited it otherwise. so I already now rectified this context towards the "meant to be" and restarted the netbox service. And that's good (working). So now I can restart the upgrade (with I minor version) from a clean situation. I'll keep you posted.

AND....
![image](https://user-images.githubusercontent.com/74967628/179937117-52a4582f-2106-4c75-bc9e-c664e24cc7ab.png)
In the upgrade documentation
...remains the fact that X.Y.Z means the new version when creating the ne symlink,
but when copying the current config files to the new version trough the symlink, X.Y.Z is actually meant 'the current version'.
So there I calls out that the documentation invokes ambiguity.
You agree?

**Reply by markkuleinio** [reply]
Well it says "latest version" and "current version". If you'd like another wording, you are free to check the contribution guidelines and propose a specific improvement in the documentation, that's how things go forward.

**Comment by samvdl** [comment]
so....
Okay, new virtual environment.... created with python 3.9
and so....

```
(venv) [ansible@netboxVM-2.9.11]$ python --version
Python 3.9.10
(venv) [ansible@netboxVM-2.9.11]$ pwd
/opt/netbox-2.9.11
(venv) [ansible@netboxVM-2.9.11]$ python netbox/manage.py migrate
Traceback (most recent call last):
  File "/opt/netbox-2.9.11/netbox/manage.py", line 8, in <module>
    from django.core.management import execute_from_command_line
ModuleNotFoundError: No module named 'django'
(venv) [ansible@netboxVM-2.9.11]$
```
but yep....
and oh by the way.... @candlerb ,  I did 
```
cd /opt/netbox-2.9.11
. venv/bin/activate
```
(I saw you post on https://groups.google.com/g/netbox-discuss/c/D30kPhx9lAw)

I'm in the right virtualenv... yes?

**Reply by markkuleinio** [reply]
> ModuleNotFoundError: No module named 'django'

Quite obviously, if you haven't ever run `upgrade.sh` (or manually done all the same operations it does)

**Reply by candlerb** [reply]
This single command:

```
PYTHON=/usr/bin/python3.9 /opt/netbox/upgrade.sh
```

will create a fresh virtualenv, populate all the dependencies, apply any outstanding migrations, collect static files - everything.

**Comment by samvdl** [comment]
So.... basically I succeeded now in the minor upgrade.
Thing is that it is imperative that /opt/netbox is a symlink to you new version folder and that you go from there.
My situation (inherited) was that the original installed version was in /opt/netbox/ itself...
And so than starting in documentation with the upgrade is conflicting.

But I must also say that in the documentation of the upgrade, X.Y.Z are used interchangeably between first the new version (for symlink creation) and than secondly the current version (for copy conf files trough symlink), which is not very consistent and was very confusing . even worse to grasp when assuming that my initial installation setup was (unknowingly) not according to initial netbox installation documentation).

...at the end I still got a 502, cause the service was not running because I needed to restart netbox and netbox-rq in the virtual environment (venv) (dumbass me!)

It was not necessary (since I was already on 2.8.9) but I thought I had to copy the services themselves from the netbox/contrib/ folder to the /etc/systemd/system/ folder as well  (reload, restart and enable them) afterwards. So I did that as well ;-)

So yes, indeed now it works. I upgraded to the latest minor version 2.11.12 and it's working

**Reply by markkuleinio** [reply]
Great!

> Thing is that it is imperative that /opt/netbox is a symlink to you new version folder

Yes, if you are following the installation option A. In general, no, /opt/netbox is not **required** to be a symlink, it is just expected that /opt/netbox can be used to find the NetBox installation root.

But anyway, good that you got through it.

**Reply by samvdl** [reply]
so there you have it ;-)....
"..in general no, except if....."  and away we are ;-)
